File: backend/layers/shared/python/lib/languages.py
"""
Language registry for VocabGym vocabulary training.

This module describes every supported language generically in a single
``LANGUAGES`` registry, plus a curated ``SUPPORTED_PAIRS`` matrix of allowed
source -> target combinations.

Historically the source language was always German and only target languages
were configurable. That is now just the *default*: German (``de``) is a normal
registry entry and the current curated pairs are all ``de -> {fr,en,es,it}``,
so behaviour is unchanged. The generic structure lets us add other source
languages later (post-paid) without touching call sites.

Backward compatibility: ``SOURCE_LANGUAGE``, ``SUPPORTED_LANGUAGES``,
``DEFAULT_TARGET_LANGUAGE``, ``get_language``, ``get_all_articles`` and
``get_article_genders`` are all preserved (derived from the registry) and behave
exactly as before.
"""

# ---------------------------------------------------------------------------
# Registry: every language described the same way.
#   register: address forms available in this language's UI (post-paid i18n).
#             'de' has informal/formal (du/Sie); romance langs tu/vous, tú/usted;
#             'en' is neutral. Only metadata for now — not yet consumed.
#   stripsDiacritics: whether answer checking may strip accents for this language
#             (true for latin-script langs here; kept explicit for future langs).
#   pollyVoices: candidate Amazon Polly voice ids (first = default).
#   latinScript: all currently supported languages are latin-script.
# ---------------------------------------------------------------------------

LANGUAGES = {
    "de": {
        "code": "de",
        "name": "Deutsch",          # German UI display name
        "nameEnglish": "German",
        "endonym": "Deutsch",
        "flag": "🇩🇪",
        "latinScript": True,
        "stripsDiacritics": True,
        "register": {"informal": "du", "formal": "Sie"},
        "pollyVoices": ["Vicki", "Daniel", "Hans", "Marlene"],
        "articles": {
            "definite": ["der", "die", "das"],
            "indefinite": ["ein", "eine", "eines", "einem", "einen", "einer"],
        },
        "articleGenders": {
            "der": "männlich (Maskulinum)",
            "die": "weiblich (Femininum) / Plural",
            "das": "sächlich (Neutrum)",
            "ein": "männlich/sächlich",
            "eine": "weiblich",
        },
    },
    "fr": {
        "code": "fr",
        "name": "Französisch",
        "nameEnglish": "French",
        "endonym": "Français",
        "flag": "🇫🇷",
        "latinScript": True,
        "stripsDiacritics": True,
        "register": {"informal": "tu", "formal": "vous"},
        "pollyVoices": ["Lea", "Celine", "Mathieu"],
        "articles": {
            "definite": ["le", "la", "les", "l'"],
            "indefinite": ["un", "une", "des", "du", "de la", "de l'", "de les"],
        },
        "articleGenders": {
            "un": "männlich (masculin)",
            "une": "weiblich (féminin)",
            "le": "männlich (masculin)",
            "la": "weiblich (féminin)",
            "les": "Plural",
            "des": "Plural (unbestimmt)",
            "l'": "männlich oder weiblich (vor Vokal)",
        },
        "examplePair": {"source": "das Haus", "target": "la maison"},
    },
    "en": {
        "code": "en",
        "name": "Englisch",
        "nameEnglish": "English",
        "endonym": "English",
        "flag": "🇬🇧",
        "latinScript": True,
        "stripsDiacritics": True,
        "register": {"neutral": "you"},
        "pollyVoices": ["Joanna", "Matthew", "Amy", "Brian"],
        "articles": {
            "definite": ["the"],
            "indefinite": ["a", "an"],
        },
        "articleGenders": {
            "a": "unbestimmt (Singular)",
            "an": "unbestimmt (vor Vokal)",
            "the": "bestimmt",
        },
        "examplePair": {"source": "das Haus", "target": "the house"},
    },
    "es": {
        "code": "es",
        "name": "Spanisch",
        "nameEnglish": "Spanish",
        "endonym": "Español",
        "flag": "🇪🇸",
        "latinScript": True,
        "stripsDiacritics": True,
        "register": {"informal": "tú", "formal": "usted"},
        "pollyVoices": ["Lucia", "Conchita", "Enrique"],
        "articles": {
            "definite": ["el", "la", "los", "las"],
            "indefinite": ["un", "una", "unos", "unas"],
        },
        "articleGenders": {
            "el": "männlich (masculino)",
            "la": "weiblich (femenino)",
            "los": "männlich Plural",
            "las": "weiblich Plural",
            "un": "männlich (unbestimmt)",
            "una": "weiblich (unbestimmt)",
        },
        "examplePair": {"source": "das Haus", "target": "la casa"},
    },
    "it": {
        "code": "it",
        "name": "Italienisch",
        "nameEnglish": "Italian",
        "endonym": "Italiano",
        "flag": "🇮🇹",
        "latinScript": True,
        "stripsDiacritics": True,
        "register": {"informal": "tu", "formal": "Lei"},
        "pollyVoices": ["Bianca", "Carla", "Giorgio"],
        "articles": {
            "definite": ["il", "lo", "la", "i", "gli", "le", "l'"],
            "indefinite": ["un", "uno", "una", "un'"],
        },
        "articleGenders": {
            "il": "männlich (maschile)",
            "lo": "männlich (vor s+Kons., z)",
            "la": "weiblich (femminile)",
            "le": "weiblich Plural",
            "un": "männlich (unbestimmt)",
            "una": "weiblich (unbestimmt)",
        },
        "examplePair": {"source": "das Haus", "target": "la casa"},
    },
}


# ---------------------------------------------------------------------------
# Curated source -> target pairs. For now only German-source pairs, so nothing
# changes for existing behaviour. promptKey selects the extraction prompt
# (post-paid: per-pair SSM prompt; today all German-source pairs share the
# existing prompt keyed 'de-*').
# ---------------------------------------------------------------------------

SUPPORTED_PAIRS = [
    {"source": "de", "target": "fr", "promptKey": "de-fr"},
    {"source": "de", "target": "en", "promptKey": "de-en"},
    {"source": "de", "target": "es", "promptKey": "de-es"},
    {"source": "de", "target": "it", "promptKey": "de-it"},
]

DEFAULT_SOURCE_LANGUAGE = "de"
DEFAULT_TARGET_LANGUAGE = "fr"


# ---------------------------------------------------------------------------
# New generic helpers
# ---------------------------------------------------------------------------

def get_registry_language(code):
    """Return the full registry entry for a language code, or None."""
    return LANGUAGES.get(code)


def get_pair(source_code, target_code):
    """Return the curated pair dict for source->target, or None if unsupported."""
    for pair in SUPPORTED_PAIRS:
        if pair["source"] == source_code and pair["target"] == target_code:
            return pair
    return None


def is_pair_supported(source_code, target_code):
    """True if the source->target combination is a curated, supported pair."""
    return get_pair(source_code, target_code) is not None


def get_articles(code):
    """Return the flat article list (definite + indefinite) for one language."""
    lang = LANGUAGES.get(code)
    if lang is None:
        return None
    return list(lang["articles"]["definite"]) + list(lang["articles"]["indefinite"])


# ---------------------------------------------------------------------------
# Backward-compatible symbols and helpers (derived from the registry).
# These preserve the historical "German source, configurable target" API.
# ---------------------------------------------------------------------------

# The historical source language (German).
SOURCE_LANGUAGE = {
    "code": LANGUAGES["de"]["code"],
    "name": LANGUAGES["de"]["name"],
    "nameEnglish": LANGUAGES["de"]["nameEnglish"],
    "articles": LANGUAGES["de"]["articles"],
    "articleGenders": LANGUAGES["de"]["articleGenders"],
}

# Historical target-language map: every language reachable as a target from the
# default German source. Currently fr, en, es, it — matching the old contents.
SUPPORTED_LANGUAGES = {
    code: LANGUAGES[code]
    for code in [
        p["target"] for p in SUPPORTED_PAIRS if p["source"] == DEFAULT_SOURCE_LANGUAGE
    ]
}


def get_language(code):
    """Return the target-language definition for a code, or None.

    Preserved for backward compatibility — only resolves target languages
    (fr/en/es/it), exactly like before. Use get_registry_language() for the
    full registry incl. the source language.
    """
    return SUPPORTED_LANGUAGES.get(code)


def get_all_articles(target_language_code):
    """Merge source (German) and target language article lists into one flat list.

    Returns None if target_language_code is not supported. Behaviour preserved.
    """
    target = SUPPORTED_LANGUAGES.get(target_language_code)
    if target is None:
        return None

    articles = []
    articles.extend(SOURCE_LANGUAGE["articles"]["definite"])
    articles.extend(SOURCE_LANGUAGE["articles"]["indefinite"])
    articles.extend(target["articles"]["definite"])
    articles.extend(target["articles"]["indefinite"])
    return articles


def get_article_genders(target_language_code):
    """Merge source (German) and target language articleGenders dicts.

    Returns None if target_language_code is not supported. Behaviour preserved.
    """
    target = SUPPORTED_LANGUAGES.get(target_language_code)
    if target is None:
        return None

    merged = {}
    merged.update(SOURCE_LANGUAGE["articleGenders"])
    merged.update(target["articleGenders"])
    return merged

File: frontend/src/utils/languages.js
/**
 * Language registry for VocabGym.
 *
 * Every language is described the same way in a single `LANGUAGES` registry,
 * plus a curated `SUPPORTED_PAIRS` matrix of allowed source -> target
 * combinations.
 *
 * Historically German was always the source and only target languages were
 * configurable. That is now just the default: German (`de`) is a normal
 * registry entry and the curated pairs are all `de -> {fr,en,es,it}`, so
 * behaviour is unchanged. The generic structure lets us add other source
 * languages later (post-paid) without touching call sites.
 *
 * Backward compatibility: `SUPPORTED_LANGUAGES` (the 4 target languages),
 * `SOURCE_LANGUAGE`, `DEFAULT_TARGET_LANGUAGE`, `getLanguage`, `getLanguageName`,
 * `getLanguageFlag` and `getAllArticleGenders` are all preserved (derived from
 * the registry) and behave exactly as before.
 */

/**
 * Full registry: every language (including the source language German).
 * - register: address forms for this language's UI (post-paid i18n metadata).
 * - stripsDiacritics: whether answer checking may strip accents for this lang.
 * - latinScript: all currently supported languages are latin-script.
 */
export const LANGUAGES = {
  de: {
    code: 'de',
    name: 'Deutsch',
    endonym: 'Deutsch',
    flag: '🇩🇪',
    latinScript: true,
    stripsDiacritics: true,
    register: { informal: 'du', formal: 'Sie' },
    articles: {
      definite: ['der', 'die', 'das'],
      indefinite: ['ein', 'eine', 'eines', 'einem', 'einen', 'einer'],
    },
    articleGenders: {
      der: 'männlich (Maskulinum)',
      die: 'weiblich (Femininum) / Plural',
      das: 'sächlich (Neutrum)',
      ein: 'männlich / sächlich',
      eine: 'weiblich',
    },
  },
  fr: {
    code: 'fr',
    name: 'Französisch',
    endonym: 'Français',
    flag: '🇫🇷',
    latinScript: true,
    stripsDiacritics: true,
    register: { informal: 'tu', formal: 'vous' },
    articles: {
      definite: ['le', 'la', 'les', "l'"],
      indefinite: ['un', 'une', 'des', 'du', 'de la', "de l'"],
    },
    articleGenders: {
      le: 'männlich (Maskulinum)',
      la: 'weiblich (Femininum)',
      les: 'Plural',
      "l'": 'männlich oder weiblich (vor Vokal)',
      un: 'männlich (Maskulinum)',
      une: 'weiblich (Femininum)',
    },
  },
  en: {
    code: 'en',
    name: 'Englisch',
    endonym: 'English',
    flag: '🇬🇧',
    latinScript: true,
    stripsDiacritics: true,
    register: { neutral: 'you' },
    articles: {
      definite: ['the'],
      indefinite: ['a', 'an'],
    },
    articleGenders: {
      the: 'bestimmter Artikel',
      a: 'unbestimmter Artikel (vor Konsonant)',
      an: 'unbestimmter Artikel (vor Vokal)',
    },
  },
  es: {
    code: 'es',
    name: 'Spanisch',
    endonym: 'Español',
    flag: '🇪🇸',
    latinScript: true,
    stripsDiacritics: true,
    register: { informal: 'tú', formal: 'usted' },
    articles: {
      definite: ['el', 'la', 'los', 'las'],
      indefinite: ['un', 'una', 'unos', 'unas'],
    },
    articleGenders: {
      el: 'männlich (Maskulinum)',
      la: 'weiblich (Femininum)',
      los: 'männlich Plural',
      las: 'weiblich Plural',
      un: 'männlich (Maskulinum)',
      una: 'weiblich (Femininum)',
      unos: 'männlich Plural',
      unas: 'weiblich Plural',
    },
  },
  it: {
    code: 'it',
    name: 'Italienisch',
    endonym: 'Italiano',
    flag: '🇮🇹',
    latinScript: true,
    stripsDiacritics: true,
    register: { informal: 'tu', formal: 'Lei' },
    articles: {
      definite: ['il', 'lo', 'la', 'i', 'gli', 'le', "l'"],
      indefinite: ['un', 'uno', 'una', "un'"],
    },
    articleGenders: {
      il: 'männlich (Maskulinum)',
      lo: 'männlich (vor s+Konsonant, z)',
      la: 'weiblich (Femininum)',
      "l'": 'männlich oder weiblich (vor Vokal)',
      i: 'männlich Plural',
      gli: 'männlich Plural (vor Vokal, s+Kons., z)',
      le: 'weiblich Plural',
      un: 'männlich (Maskulinum)',
      uno: 'männlich (vor s+Konsonant, z)',
      una: 'weiblich (Femininum)',
      "un'": 'weiblich (vor Vokal)',
    },
  },
}

/**
 * Curated source -> target pairs. For now only German-source pairs, so nothing
 * changes for existing behaviour.
 */
export const SUPPORTED_PAIRS = [
  { source: 'de', target: 'fr', promptKey: 'de-fr' },
  { source: 'de', target: 'en', promptKey: 'de-en' },
  { source: 'de', target: 'es', promptKey: 'de-es' },
  { source: 'de', target: 'it', promptKey: 'de-it' },
]

export const DEFAULT_SOURCE_LANGUAGE = 'de'

/**
 * Default target language code.
 */
export const DEFAULT_TARGET_LANGUAGE = 'fr'

/**
 * Source language (German) configuration — derived from the registry.
 * Preserved for backward compatibility.
 */
export const SOURCE_LANGUAGE = {
  code: LANGUAGES.de.code,
  name: LANGUAGES.de.name,
  flag: LANGUAGES.de.flag,
  articles: LANGUAGES.de.articles,
  articleGenders: LANGUAGES.de.articleGenders,
}

/**
 * Supported target languages (fr, en, es, it) — derived from the pair matrix
 * for the default German source. Preserved for backward compatibility: this
 * intentionally does NOT include the source language.
 */
export const SUPPORTED_LANGUAGES = Object.fromEntries(
  SUPPORTED_PAIRS
    .filter((p) => p.source === DEFAULT_SOURCE_LANGUAGE)
    .map((p) => [p.target, LANGUAGES[p.target]])
)

// ---------------------------------------------------------------------------
// New generic helpers
// ---------------------------------------------------------------------------

/**
 * Return the full registry entry for a language code (incl. the source lang).
 * @param {string} code
 * @returns {object|undefined}
 */
export function getRegistryLanguage(code) {
  return LANGUAGES[code]
}

/**
 * Return the curated pair for source->target, or undefined if unsupported.
 * @param {string} source
 * @param {string} target
 * @returns {object|undefined}
 */
export function getPair(source, target) {
  return SUPPORTED_PAIRS.find((p) => p.source === source && p.target === target)
}

/**
 * True if the source->target combination is a curated, supported pair.
 * @param {string} source
 * @param {string} target
 * @returns {boolean}
 */
export function isPairSupported(source, target) {
  return getPair(source, target) !== undefined
}

// ---------------------------------------------------------------------------
// Backward-compatible helpers (unchanged behaviour)
// ---------------------------------------------------------------------------

/**
 * Get target-language configuration by code.
 * @param {string} code - Language code (e.g. 'fr', 'en', 'es', 'it')
 * @returns {object|undefined} Language configuration object
 */
export function getLanguage(code) {
  return SUPPORTED_LANGUAGES[code]
}

/**
 * Get the German UI name for a language code. Resolves against the full
 * registry so source-language codes (e.g. 'de') also resolve.
 * @param {string} code - Language code
 * @returns {string} German name of the language, or the code itself if unknown
 */
export function getLanguageName(code) {
  return LANGUAGES[code]?.name || code
}

/**
 * Get the flag emoji for a language code (resolves against the full registry).
 * @param {string} code - Language code
 * @returns {string} Flag emoji, or empty string if unknown
 */
export function getLanguageFlag(code) {
  return LANGUAGES[code]?.flag || ''
}

/**
 * Get all article-to-gender mappings for both source (German) and target language.
 * @param {string} targetCode - Target language code
 * @returns {object} Merged object of article -> German gender explanation
 */
export function getAllArticleGenders(targetCode) {
  const targetLang = SUPPORTED_LANGUAGES[targetCode]
  const targetGenders = targetLang?.articleGenders || {}
  return { ...SOURCE_LANGUAGE.articleGenders, ...targetGenders }
}
